pkg/trie/triedb/triedb.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package triedb
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/trie"
+	nibbles "github.com/ChainSafe/gossamer/pkg/trie/codec"
+	"github.com/ChainSafe/gossamer/pkg/trie/db"
+
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb/codec"
+)
+
+var ErrIncompleteDB = errors.New("incomplete database")
+
+// TrieDB is a DB-backed patricia merkle trie implementation
+// using lazy loading to fetch nodes
+type TrieDB struct {
+	rootHash common.Hash
+	db       db.DBGetter
+}
+
+// NewTrieDB creates a new TrieDB using the given root and db
+func NewTrieDB(rootHash common.Hash, db db.DBGetter) *TrieDB {
+	return &TrieDB{
+		rootHash: rootHash,
+		db:       db,
+	}
+}
+
+// Hash returns the hashed root of the trie.
+func (t *TrieDB) Hash() (common.Hash, error) {
+	// This is trivial since it is a read only trie, but will change when we
+	// support writes
+	return t.rootHash, nil
+}
+
+// MustHash returns the hashed root of the trie.
+// It panics if it fails to hash the root node.
+func (t *TrieDB) MustHash() common.Hash {
+	h, err := t.Hash()
+	if err != nil {
+		panic(err)
+	}
+
+	return h
+}
+
+// Get returns the value in the node of the trie
+// which matches its key with the key given.
+// Note the key argument is given in little Endian format.
+func (t *TrieDB) Get(key []byte) []byte {
+	val, err := t.lookup(key)
+	if err != nil {
+		// TODO: do we have to do anything else? maybe change the signature
+		// to return an error?
+		return nil
```

Comment:
This is up to you, I am ok with just returning the error

Reply:
We can address it in the future because I'll need to change the interface and the in-memory trie too

Reply:
Returning `nil` makes sense to me.